include/fiction/algorithms/simulation/sidb/operational_domain.hpp

```diff
@@ -624,27 +712,60 @@ class operational_domain_impl
 
         mockturtle::stopwatch stop{stats.time_total};
 
-        // step points are analyzed sequentially because the CDS is updated for each step point, so parallelizing may
-        // result in unexpected behavior.
-        std::for_each(x_indices.cbegin(), x_indices.cend(),
-                      [this, &lyt](const auto x)
-                      {
-                          std::for_each(y_indices.cbegin(), y_indices.cend(),
-                                        [this, &lyt, x](const auto y) { is_step_point_suitable(lyt, {x, y}); });
-                      });
+        const auto all_step_points = cartesian_combinations(indices);
+
+        // calculate the size of each slice
+        const auto slice_size = (all_step_points.size() + num_threads - 1) / num_threads;
+
+        std::vector<std::thread> threads{};
+        threads.reserve(num_threads);
+
+        // launch threads, each with its own slice of random step points
+        for (auto i = 0ul; i < num_threads; ++i)
+        {
+            const auto start = i * slice_size;
+            const auto end   = std::min(start + slice_size, all_step_points.size());
+
+            if (start >= end)
+            {
+                break;  // no more work to distribute
+            }
+
+            threads.emplace_back(
+                [this, &lyt, start, end, &all_step_points]
+                {
+                    for (auto it = all_step_points.cbegin() + static_cast<int64_t>(start);
+                         it != all_step_points.cbegin() + static_cast<int64_t>(end); ++it)
+                    {
+                        is_step_point_suitable(lyt, step_point{*it});
```

Comment:
```suggestion
                        is_step_point_suitable(lyt, *it);
```
@marcelwa I think I did this wrong here.

Reply:
Good catch. We can avoid a copy-construction here.

Reply:
@marcelwa ah, right! would it be worth renaming ``all_step_points`` to avoid this kind of confusion? 😅

Reply:
Absolutely! We had it named `all_index_combinations` in `grid_search`. I'll adjust it accordingly.